mcs/class/Mono.Profiler.Log/Mono.Profiler.Log/LogProcessor.cs

```diff
@@ -0,0 +1,619 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Threading;
+
+namespace Mono.Profiler.Log {
+
+	public sealed class LogProcessor {
+
+		public LogReader Reader { get; }
+
+		public LogEventVisitor Visitor { get; }
+
+		readonly LogEventPool _pool = new LogEventPool ();
+
+		LogStreamHeader _streamHeader;
+
+		LogBufferHeader _bufferHeader;
+
+		ulong _time;
+
+		bool _done;
+
+		public LogProcessor (LogReader reader, LogEventVisitor visitor)
+		{
+			if (reader == null)
+				throw new ArgumentNullException (nameof (reader));
+
+			if (visitor == null)
+				throw new ArgumentNullException (nameof (visitor));
+
+			Reader = reader;
+			Visitor = visitor;
+		}
+
+		public void Process ()
+		{
+			Process (CancellationToken.None);
+		}
+
+		public void Process (CancellationToken token)
+		{
+			if (_done)
+				throw new InvalidOperationException ("This processor cannot be reused.");
+
+			_streamHeader = new LogStreamHeader (Reader);
+
+			var events = new List<LogEvent> (Environment.ProcessorCount * 1000);
+
+			void ProcessEvents ()
+			{
+				foreach (var ev in events.OrderBy (x => x.Timestamp)) {
+					token.ThrowIfCancellationRequested ();
+
+					ev.Accept (Visitor);
+					_pool.Return (ev);
+				}
+
```

Comment:
Another optimisation we recently added to the Xamarin Profiler was about reading the file in one thread and processing it in another (https://github.com/xamarin/profiler/blob/master/src/XamarinProfiler.Core/Platforms/Mono/MlpdReader.cs#L164), so that we can send to the UI live timing information, as well as perf counters, for instance, while we process heavier events (snapshots, etc) on a separate thread. For that, we've had to make the whole API async -> https://github.com/xamarin/profiler/tree/master/src/XamarinProfiler.Core/Platforms/Mono so it would be good if this API had that optimisation in. If not, we'll get back to having the UIs look stuck while processing huge MLPDs